c2rust-analyze/src/rewrite/shim.rs

```diff
@@ -0,0 +1,240 @@
+///
+/// The criteria we look for are:
+/// * The callee must be a local function.
+/// * The callee must have at least one non-`FIXED` pointer type in its signature.
+/// * The callee must not be a trait method.  Adding shims for trait methods is more complex than
+///   handling free functions or inherent methods.
+pub fn gen_shim_call_rewrites<'tcx>(
+    gacx: &GlobalAnalysisCtxt<'tcx>,
+    gasn: &GlobalAssignment,
+) -> (Vec<(Span, Rewrite)>, HashSet<DefId>) {
+    let tcx = gacx.tcx;
+
+    let mut rewrites = Vec::new();
+    let mut mentioned_fns = HashSet::new();
+
+    for &failed_def_id in gacx.fns_failed.keys() {
+        let failed_def_id = match failed_def_id.as_local() {
+            Some(x) => x,
+            None => continue,
+        };
+        let hir_body_id = tcx.hir().body_owned_by(failed_def_id);
+        let hir = tcx.hir().body(hir_body_id);
+        let typeck_results = tcx.typeck_body(hir_body_id);
+        let mut v = ShimCallVisitor {
+            gacx,
+            gasn,
+            typeck_results,
+            rewrites,
+            mentioned_fns,
+        };
+        intravisit::Visitor::visit_body(&mut v, hir);
+
+        rewrites = v.rewrites;
+        mentioned_fns = v.mentioned_fns;
+    }
+
+    (rewrites, mentioned_fns)
+}
+
+/// Convert an `LTy` to a pair of `TypeDesc`s, one computed normally and one with `FIXED` added.
+/// Returns `None` if the input `LTy` already has `FIXED` set.
+fn lty_to_desc_pair<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    gasn: &GlobalAssignment,
+    lty: LTy<'tcx>,
+) -> Option<(TypeDesc<'tcx>, TypeDesc<'tcx>)> {
+    let ptr = lty.label;
+    if ptr.is_none() || gasn.flags[ptr].contains(FlagSet::FIXED) {
+        return None;
+    }
+
+    let desc = type_desc::perms_to_desc(lty.ty, gasn.perms[ptr], gasn.flags[ptr]);
+    let fixed_desc = type_desc::perms_to_desc_with_pointee(
+        tcx,
+        desc.pointee_ty,
+        lty.ty,
+        gasn.perms[ptr],
+        gasn.flags[ptr] | FlagSet::FIXED,
+    );
+    Some((desc, fixed_desc))
+}
+
+pub fn gen_shim_definition_rewrite<'tcx>(
+    gacx: &GlobalAnalysisCtxt<'tcx>,
+    gasn: &GlobalAssignment,
+    def_id: DefId,
+) -> (Span, Rewrite) {
+    let tcx = gacx.tcx;
+
+    let owner_node = tcx.hir().expect_owner(def_id.as_local().unwrap());
+    let insert_span = owner_node.span().shrink_to_hi();
+
+    let fn_decl = owner_node.fn_decl().unwrap();
+    let arg_tys = fn_decl
+        .inputs
+        .iter()
+        .map(|ty| Rewrite::Extract(ty.span))
+        .collect::<Vec<_>>();
+    let return_ty = match fn_decl.output {
+        FnRetTy::DefaultReturn(..) => None,
+        FnRetTy::Return(ty) => Some(Box::new(Rewrite::Extract(ty.span))),
+    };
+
+    // `def_id` should always refer to a rewritten function, and all rewritten functions have
+    // valid `fn_sigs` entries.
+    let lsig = gacx
+        .fn_sigs
+        .get(&def_id)
+        .unwrap_or_else(|| panic!("missing lsig for {:?}", def_id));
```

Comment:
nit: We have a bunch of these, and I think we can just use `[...]` indexing here vs. custom panic messages.

```suggestion
    let lsig = gacx.fn_sigs[&def_id];
```